### PR TITLE
Append empty docker credentials only if ServicePrincipal is provided

### LINKED ISSUE #82043: Unable to download from anonymous Azure Container Registry with containerd
**What happened**:

Problem is that
 
https://github.com/kubernetes/kubernetes/blob/bfd8610ddafb2fcb2cb330187c8248777ea0b022/pkg/credentialprovider/azure/azure_credentials.go#L209
 
Seems like an empty username and password for AuthConfig is update for *azurecr* instead do not have an entry to keep a nil auth context.
 
But according to the code https://github.com/kubernetes/kubernetes/blob/bfd8610ddafb2fcb2cb330187c8248777ea0b022/pkg/kubelet/kuberuntime/kuberuntime_image.go#L44 this sends an empty AuthConfig since we update the structure with empty creds, which containerd doesn’t like
 
So instead of sending this
 
`imageRef, err := m.imageService.PullImage(imgSpec, nil, podSandboxConfig)`
 
 
we end up sending this
 
`imageRef, err := m.imageService.PullImage(imgSpec, auth, podSandboxConfig)`
 
with an empty auth context.
 
https://github.com/containerd/cri/blob/6dc2a874470f7489217a7c757da258356b92bfdb/pkg/server/image_pull.go#L190
 
which matches this error
 
```
Aug 27 16:39:37 containerd[4953]: time="2019-08-27T16:39:37.609386516Z" level=error msg="PullImage "repos.azurecr.io/mirror/download-amd64:v1" failed" error="failed to resolve image "repos.azurecr.io/mirror/download-amd64:v1": no available registry endpoint: invalid auth config"
```
 
**What you expected to happen**:

Containerd should successfully download from anonymous ACR

**How to reproduce it (as minimally and precisely as possible)**:

use kubelet to download pod from anonymous ACR with containerd CRI

**Anything else we need to know?**:

ideally the check should be made if the repo is anonymous, do not add to the config, which is done in MSI case. Worst case atleast add this to Service Principal case, but that would still be a hack
 
``` golang
else {
               // Add our entry for each of the supported container registry URLs
               for _, url := range containerRegistryUrls {
                       cred := &credentialprovider.DockerConfigEntry{
                               Username: a.config.AADClientID,
                               Password: a.config.AADClientSecret,
                               Email:    dummyRegistryEmail,
                       }
                       cfg[url] = *cred
               }
               // add ACR anonymous repo support: use empty username and password for anonymous access
                cfg["*.azurecr.*"] = credentialprovider.DockerConfigEntry{
                        Username: "",
                       Password: "",
                       Email:    dummyRegistryEmail,
                }
 
        }
 ```
 
 **Environment**:
- Kubernetes version (use `kubectl version`): 1.14.6
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`): ubuntu
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by awesomenix**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:

- In case of service principal scenario credentials are blindly added without anonymous check
- Add empty credentials as well in that scenario

**Which issue(s) this PR fixes**:

Fixes #82043

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Append empty docker credentials only if ServicePrincipal is provided for ACR
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/credentialprovider/azure/azure_credentials.go [hunk 1]
```diff
@@ -203,14 +204,14 @@ func (a *acrProvider) Provide(image string) credentialprovider.DockerConfig {
...
 		}
+		// add ACR anonymous repo support: use empty username and password for anonymous access
+		cfg["*.azurecr.*"] = credentialprovider.DockerConfigEntry{
+			Username: "",
+			Password: "",
+			Email:    dummyRegistryEmail,
+		}
 	}
 
-	// add ACR anonymous repo support: use empty username and password for anonymous access
```
Comment (andyzhangx) [inline pkg/credentialprovider/azure/azure_credentials.go]: why remove this? when user use `UseManagedIdentityExtension`,  also needs to support ACR anonymous.
Original code will always have `empty username and password for anonymous access`, will that solve your issue?

**Review by feiskyer — COMMENTED** [review]
I don't think this would fix the issue. We don't know whether ACR is anonymous or not.

@awesomenix Have you checked with ACR team for supporting empty auth context?

File: pkg/credentialprovider/azure/azure_credentials.go (hunk 1 above)
Comment (feiskyer) [reply pkg/credentialprovider/azure/azure_credentials.go]: The provided credential may not be authz to the ACR, so empty auth context is required for both MSI and SPN.

**Comment by andyzhangx** [comment]
> I don't think this would fix the issue. We don't know whether ACR is anonymous or not.
> 
> @awesomenix Have you checked with ACR team for supporting empty auth context?

`empty auth context` is we provide empty username and password, and thus we could access the anonymous ACR repo, that's guaranteed by ACR team, they came out the solution when I made the fix last time.

For MSI case, if we don't provide empty username and password(this PR removes the code for MSI), kubelet will only use MSI to communicate with ACR repo, so if that's an anonymous ACR repo, this will not work. 
add @sivagms for any comments. Thanks.

**Comment by andyzhangx** [comment]
> Yes, even if MSI is used, you may need to access an ACR registry with anonymous access enabled AND the MSI may not have/need access to this registry. In such case, empty credential will work.
> /cc @sajayantony

Thanks for clarification, so this PR will have regression, there is another way to fix this issue: https://github.com/containerd/cri/pull/1250, we'd better use that fix

/hold

**Pull request closed without merging**